tokio-util/src/sync/mpsc.rs

```diff
@@ -1,221 +1,282 @@
+            if let Some(result) = result {
+                return result;
+            }
         }
-        result
     }
 
-    /// Check whether the channel is ready to send more messages now.
+    /// Sends an item to the channel.
     ///
-    /// If this method returns `true`, then `start_send` is guaranteed to not
-    /// panic.
+    /// Before calling `start_send`, `poll_reserve` must be called with a successful return
+    /// value of `Poll::Ready(Ok(()))`.
     ///
-    /// If the channel is closed, this method returns `true`.
-    pub fn is_ready(&self) -> bool {
-        !self.is_sending
-    }
+    /// # Errors
+    ///
+    /// If the channel is closed, an error will be returned.  This is a permanent state.
+    ///
+    /// # Panics
+    ///
+    /// If `poll_reserve` was not successfully called prior to calling `start_send`, then this method
+    /// will panic.
+    pub fn start_send(&mut self, value: T) -> Result<(), PollSendError<T>> {
+        let (result, next_state) = match self.take_state() {
+            State::Idle(_) | State::Acquiring => {
+                panic!("`start_send` called without first calling `poll_reserve`")
+            }
+            // We have a permit to send our item, so go ahead, which gets us our sender back.
+            State::ReadyToSend(permit) => (Ok(()), State::Idle(permit.send(value))),
+            // We're closed, either by choice or because the underlying sender was closed.
+            State::Closed => (Err(PollSendError(Some(value))), State::Closed),
+        };
 
-    /// Check whether the channel has been closed.
-    pub fn is_closed(&self) -> bool {
-        match &self.sender {
-            Some(sender) => sender.is_closed(),
-            None => true,
-        }
+        // Handle deferred closing if `close` was called between `poll_reserve` and `start_send`.
+        self.state = if self.sender.is_some() {
+            next_state
+        } else {
+            State::Closed
+        };
+        result
     }
 
-    /// Clone the underlying `Sender`.
+    /// Checks whether this sender is been closed.
     ///
-    /// If this method returns `None`, then the channel is closed. (But it is
-    /// not guaranteed to return `None` if the channel is closed.)
-    pub fn clone_inner(&self) -> Option<Sender<T>> {
-        self.sender.as_ref().map(|sender| (&**sender).clone())
+    /// The underlying channel that this sender was wrapping may still be open.
+    pub fn is_closed(&self) -> bool {
+        matches!(self.state, State::Closed) || self.sender.is_none()
     }
 
-    /// Access the underlying `Sender`.
+    /// Gets a reference to the underlying `Sender`.
     ///
-    /// If this method returns `None`, then the channel is closed. (But it is
-    /// not guaranteed to return `None` if the channel is closed.)
-    pub fn inner_ref(&self) -> Option<&Sender<T>> {
-        self.sender.as_deref()
+    /// If the channel is closed, `None` is returned.
```

Comment:
This might return `Some` even if the channel is closed.

Reply:
I guess the better comment would be "If this sender is closed, `None` is returned.`"

The implication is that we drop our references to the underlying channel sender when asked to close.

Reply:
There is a comment similar to that on the existing implementation.

Reply:
That situation seems, potentially, very normal: `Receiver` closes itself, a bunch of `Sender`s still exist and haven't yet tried to send, or check the channel status, to figure out the channel is closed and that they can't send anything else.

If a user calls `poll_reserve` and we encounter an error reserving the permit, we mark _ourselves_ as closed, so our close state is eventually consistent with the underlying channel.

Reply:
I don't think there's anything wrong with how its implemented here - I just think the comment should be updated.

Reply:
Sorry, looks like I responded to the top-level comment twice from a stale Chrome tab or something.

I'll update the comment. 👍🏻

Reply:
Addressed in eeb4abdf.